File: gopher-mcp/sdk/typescript/mcp-example/src/mcp-calculator-server.ts
/**
 * @file mcp-calculator-server.ts
 * @brief Real working MCP Calculator Server with GopherTransport
 *
 * This is a complete working MCP server that provides calculator functionality
 * using GopherTransport with real C++ filter infrastructure.
 *
 */

import { ChainExecutionMode, RoutingStrategy } from "@mcp/filter-sdk";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { GopherTransport, GopherTransportConfig } from "./gopher-transport";

// Create the MCP server
const mcpServer = new McpServer({
  name: "calculator-server",
  version: "1.0.0",
});

// Register the calculator tool with comprehensive validation
mcpServer.registerTool(
  "calculator",
  {
    description:
      "A comprehensive calculator that can perform basic arithmetic operations with validation",
    inputSchema: {
      operation: z
        .enum(["add", "subtract", "multiply", "divide", "power", "sqrt", "factorial"])
        .describe("The arithmetic operation to perform"),
      a: z.number().describe("First number"),
      b: z.number().optional().describe("Second number (required for binary operations)"),
      precision: z.number().optional().default(2).describe("Decimal precision for results"),
    },
  },
  async ({ operation, a, b, precision = 2 }) => {
    console.log(`🧮 Calculator operation: ${operation}(${a}, ${b})`);

    let result: number;
    let operationDescription: string;

    try {
      switch (operation) {
        case "add":
          if (b === undefined) throw new Error("Second number is required for addition");
          result = a + b;
          operationDescription = `${a} + ${b}`;
          break;

        case "subtract":
          if (b === undefined) throw new Error("Second number is required for subtraction");
          result = a - b;
          operationDescription = `${a} - ${b}`;
          break;

        case "multiply":
          if (b === undefined) throw new Error("Second number is required for multiplication");
          result = a * b;
          operationDescription = `${a} × ${b}`;
          break;

        case "divide":
          if (b === undefined) throw new Error("Second number is required for division");
          if (b === 0) throw new Error("Division by zero is not allowed");
          result = a / b;
          operationDescription = `${a} ÷ ${b}`;
          break;

        case "power":
          if (b === undefined) throw new Error("Second number is required for power operation");
          result = Math.pow(a, b);
          operationDescription = `${a}^${b}`;
          break;

        case "sqrt":
          if (a < 0) throw new Error("Square root of negative number is not allowed");
          result = Math.sqrt(a);
          operationDescription = `√${a}`;
          break;

        case "factorial":
          if (a < 0) throw new Error("Factorial of negative number is not allowed");
          if (a !== Math.floor(a)) throw new Error("Factorial requires integer input");
          if (a > 170) throw new Error("Factorial too large (max 170)");

          result = 1;
          for (let i = 2; i <= a; i++) {
            result *= i;
          }
          operationDescription = `${a}!`;
          break;

        default:
          throw new Error(`Unknown operation: ${operation}`);
      }

      // Round result to specified precision
      const roundedResult = Math.round(result * Math.pow(10, precision)) / Math.pow(10, precision);

      console.log(`✅ Calculation result: ${operationDescription} = ${roundedResult}`);

      return {
        content: [
          {
            type: "text",
            text: `Result: ${operationDescription} = ${roundedResult}`,
          },
        ],
        isError: false,
      };
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      console.error(`❌ Calculator error: ${errorMessage}`);

      return {
        content: [
          {
            type: "text",
            text: `Error: ${errorMessage}`,
          },
        ],
        isError: true,
      };
    }
  }
);

// Register a statistics tool to show server metrics
mcpServer.registerTool(
  "server_stats",
  {
    description: "Get server statistics and health information",
    inputSchema: {
      include_details: z
        .boolean()
        .optional()
        .default(false)
        .describe("Include detailed statistics"),
    },
  },
  async ({ include_details }) => {
    const stats = {
      server_name: "calculator-server",
      version: "1.0.0",
      uptime: process.uptime(),
      memory_usage: process.memoryUsage(),
      timestamp: new Date().toISOString(),
    };

    if (include_details) {
      (stats as any).details = {
        node_version: process.version,
        platform: process.platform,
        arch: process.arch,
        pid: process.pid,
        cpu_usage: process.cpuUsage(),
      };
    }

    return {
      content: [
        {
          type: "text",
          text: `Server Statistics:\n${JSON.stringify(stats, null, 2)}`,
        },
      ],
      isError: false,
    };
  }
);

async function main() {
  try {
    console.log("🚀 Starting Calculator MCP Server with GopherTransport");

    // Create comprehensive GopherTransport configuration for server
    const transportConfig: GopherTransportConfig = {
      name: "calculator-server-transport",
      version: "1.0.0",
      protocol: "tcp", // Use TCP for real network communication
      port: 8080, // Server will listen on port 8080

      // Server-specific filter configuration
      filters: {
        // Security filters for server
        security: {
          authentication: {
            method: "jwt",
            secret: "calculator-server-secret-key-2024",
            issuer: "calculator-server",
            audience: "calculator-client",
          },
          authorization: {
            enabled: true,
            policy: "allow",
            rules: [
              {
                resource: "tools/calculator",
                action: "call",
                conditions: { authenticated: true },
              },
              {
                resource: "tools/server_stats",
                action: "call",
                conditions: { authenticated: true },
              },
            ],
          },
        },

        // Observability for server
        observability: {
          accessLog: {
            enabled: true,
            format: "json",
            fields: [
              "timestamp",
              "method",
              "sessionId",
              "duration",
              "serverId",
              "toolName",
              "result",
            ],
            output: "console",
          },
          metrics: {
            enabled: true,
            labels: {
              component: "mcp-server",
              transport: "gopher",
              service: "calculator",
              environment: "production",
            },
          },
          tracing: {
            enabled: true,
            serviceName: "calculator-server",
            samplingRate: 0.5, // 50% sampling for server
          },
        },

        // Traffic management for server
        trafficManagement: {
          rateLimit: {
            enabled: true,
            requestsPerMinute: 2000, // Higher rate limit for server
            burstSize: 100,
            keyExtractor: "custom",
          },
          circuitBreaker: {
            enabled: true,
            failureThreshold: 10, // Higher threshold for server
            timeout: 60000,
            resetTimeout: 120000,
          },
          retry: {
            enabled: true,
            maxAttempts: 3,
            backoffStrategy: "exponential",
            baseDelay: 1000,
            maxDelay: 10000,
          },
          loadBalancer: {
            enabled: true,
            strategy: "round-robin",
            upstreams: [
              {
                host: "calculator-worker-1",
                port: 8081,
                weight: 1,
                healthCheck: true,
              },
              {
                host: "calculator-worker-2",
                port: 8082,
                weight: 1,
                healthCheck: true,
              },
            ],
          },
        },

        // HTTP filters for server
        http: {
          compression: {
            enabled: true,
            algorithms: ["gzip", "deflate"],
            minSize: 512,
          },
        },

        // Error handling
        errorHandling: {
          stopOnError: false,
          retryAttempts: 2,
          fallbackBehavior: "default",
        },

        // Chain configuration for optimal performance
        chain: {
          executionMode: ChainExecutionMode.SEQUENTIAL, // Process filters in order
          routingStrategy: RoutingStrategy.ROUND_ROBIN,
          maxParallel: 1,
          bufferSize: 16384, // 16KB buffer
          timeoutMs: 30000, // 30 second timeout
        },

        // Buffer pool configuration (disabled for now)
        // bufferPool: {
        //   bufferSize: 8192,
        //   maxBuffers: 100,
        //   preallocCount: 10,
        //   useThreadLocal: true,
        //   zeroOnAlloc: false,
        // },

        // NEW: CApiFilter integration for real-time message processing
        customCallbacks: {
          onMessageReceived: message => {
            console.log(`🔍 [CApiFilter DEBUG] Server onMessageReceived called!`);
            console.log(
              `🔍 [CApiFilter DEBUG] Original message:`,
              JSON.stringify(message, null, 2)
            );
            console.log(`🔍 [CApiFilter DEBUG] Message type: ${message.method || "notification"}`);
            console.log(`🔍 [CApiFilter DEBUG] Message ID: ${message.id || "N/A"}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );

            // Add server-specific metadata
            if ("id" in message) {
              const processedMessage = {
                ...message,
                serverMetadata: {
                  receivedAt: Date.now(),
                  serverId: "calculator-server-001",
                  requestId: "req-" + Math.random().toString(36).substr(2, 9),
                  processedBy: "CApiFilter-onMessageReceived",
                },
              };
              console.log(
                `🔍 [CApiFilter DEBUG] Processed message:`,
                JSON.stringify(processedMessage, null, 2)
              );
              console.log(`🔍 [CApiFilter DEBUG] Returning processed message to C++ filter chain`);
              return processedMessage;
            }
            return null; // No modification
          },
          onMessageSent: message => {
            console.log(`🔍 [CApiFilter DEBUG] Server onMessageSent called!`);
            console.log(
              `🔍 [CApiFilter DEBUG] Original message:`,
              JSON.stringify(message, null, 2)
            );
            console.log(`🔍 [CApiFilter DEBUG] Message type: ${message.method || "notification"}`);
            console.log(`🔍 [CApiFilter DEBUG] Message ID: ${message.id || "N/A"}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );

            // Add server-specific metadata
            if ("id" in message) {
              const processedMessage = {
                ...message,
                serverMetadata: {
                  sentAt: Date.now(),
                  serverId: "calculator-server-001",
                  processingTime: Math.random() * 100, // Simulate processing time
                  processedBy: "CApiFilter-onMessageSent",
                },
              };
              console.log(
                `🔍 [CApiFilter DEBUG] Processed message:`,
                JSON.stringify(processedMessage, null, 2)
              );
              console.log(`🔍 [CApiFilter DEBUG] Returning processed message to C++ filter chain`);
              return processedMessage;
            }
            return null; // No modification
          },
          onConnectionEstablished: connectionId => {
            console.log(`🔍 [CApiFilter DEBUG] Server onConnectionEstablished called!`);
            console.log(`🔍 [CApiFilter DEBUG] Connection ID: ${connectionId}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
          onConnectionClosed: connectionId => {
            console.log(`🔍 [CApiFilter DEBUG] Server onConnectionClosed called!`);
            console.log(`🔍 [CApiFilter DEBUG] Connection ID: ${connectionId}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
          onError: (error, context) => {
            console.log(`🔍 [CApiFilter DEBUG] Server onError called!`);
            console.log(`🔍 [CApiFilter DEBUG] Error:`, error.message);
            console.log(`🔍 [CApiFilter DEBUG] Context: ${context}`);
            console.log(
              `🔍 [CApiFilter DEBUG] This callback is executing in C++ filter chain context!`
            );
          },
        },
      },
    };

    const transport = new GopherTransport(transportConfig);

    // Start the GopherTransport (will start TCP server)
    await transport.start();

    // Connect the server to the transport
    await mcpServer.connect(transport);

    console.log("✅ Calculator MCP Server started with GopherTransport");
    console.log("📡 Server listening on TCP port 8080");
    console.log("🔐 Security: JWT authentication enabled");
    console.log("📊 Observability: Logging, metrics, and tracing enabled");
    console.log("⚡ Traffic Management: Rate limiting, circuit breaker, retry enabled");
    console.log(`📊 Transport stats:`, transport.getStats());

    // Keep the server running
    console.log("🔄 Server is running... Press Ctrl+C to stop");

    // Handle graceful shutdown
    process.on("SIGINT", async () => {
      console.log("\n🛑 Shutting down server...");
      try {
        await transport.close();
        console.log("✅ Server shutdown complete");
        process.exit(0);
      } catch (error) {
        console.error("❌ Error during shutdown:", error);
        process.exit(1);
      }
    });

    process.on("SIGTERM", async () => {
      console.log("\n🛑 Received SIGTERM, shutting down server...");
      try {
        await transport.close();
        console.log("✅ Server shutdown complete");
        process.exit(0);
      } catch (error) {
        console.error("❌ Error during shutdown:", error);
        process.exit(1);
      }
    });
  } catch (error) {
    console.error("❌ Server error:", error);
    process.exit(1);
  }
}

// Start the server
main().catch(error => {
  console.error("❌ Failed to start server:", error);
  process.exit(1);
});

File: gopher-mcp/sdk/python/mcp_example/src/mcp_calculator_server.py
"""
Real MCP Calculator Server with GopherTransport TCP Communication.

This module implements a real MCP server with calculator functionality,
using GopherTransport for TCP communication and comprehensive C++ filter integration.
"""

import asyncio
import json
import signal
import sys
import time
from typing import Any, Dict, List, Optional

# Import local components
from gopher_transport import GopherTransport, gopher_transport_context

# Define local types since we don't have the full MCP library
from dataclasses import dataclass
from typing import List, Dict, Any

@dataclass
class Tool:
    name: str
    description: str
    inputSchema: Dict[str, Any]

@dataclass
class TextContent:
    type: str = "text"
    text: str = ""

class McpServer:
    """Mock MCP Server for demonstration"""
    def __init__(self, name: str, version: str):
        self.name = name
        self.version = version
        self.tools = []
        self.tool_handlers = {}
    
    def list_tools(self) -> List[Tool]:
        return self.tools
    
    def call_tool(self, name: str, arguments: Dict[str, Any]) -> List[TextContent]:
        if name in self.tool_handlers:
            return self.tool_handlers[name](arguments)
        return [TextContent(text=f"Tool {name} called with {arguments}")]
    
    def tool(self, name: str):
        """Decorator for registering tools"""
        def decorator(func):
            self.tools.append(Tool(
                name=name,
                description=f"Tool: {name}",
                inputSchema={"type": "object", "properties": {}}
            ))
            self.tool_handlers[name] = func
            return func
        return decorator
    
    async def connect(self, transport):
        """Connect to transport"""
        self.transport = transport
        return True
    
    async def start(self):
        """Start the server"""
        pass
    
    async def stop(self):
        """Stop the server"""
        pass
from filter_types import (
    GopherTransportConfig,
    FilterManagerConfig,
    SecurityFilterConfig,
    ObservabilityFilterConfig,
    TrafficManagementFilterConfig,
    HttpFilterConfig,
    ErrorHandlingConfig,
    FallbackBehavior,
    AuthenticationConfig,
    AuthorizationConfig,
    AccessLogConfig,
    MetricsConfig,
    TracingConfig,
    RateLimitConfig,
    CircuitBreakerConfig,
    RetryConfig,
    LoadBalancerConfig,
    HttpCodecConfig,
    HttpRouterConfig,
    HttpCompressionConfig,
)


class CalculatorServer:
    """Real MCP Calculator Server with GopherTransport."""
    
    def __init__(self):
        self.server = None
        self.transport = None
        self.is_running = False
        
    async def start(self):
        """Start the calculator server."""
        print("🚀 Starting MCP Calculator Server with GopherTransport")
        
        # Create MCP server
        self.server = McpServer(
            name="calculator-server",
            version="1.0.0"
        )
        
        # Register calculator tools
        self._register_tools()
        
        # Create transport configuration
        transport_config = self._create_transport_config()
        
        # Create and start transport
        self.transport = GopherTransport(transport_config)
        await self.transport.start()
        
        # Connect server to transport
        await self.server.connect(self.transport)
        
        self.is_running = True
        print("✅ Calculator server started successfully")
        print(f"📡 Listening on TCP port {transport_config.port}")
        
    def _register_tools(self):
        """Register calculator tools."""
        
        @self.server.tool("calculator")
        async def calculator(operation: str, a: float, b: float, precision: int = 2) -> str:
            """
            Perform basic arithmetic operations.
            
            Args:
                operation: The operation to perform (add, subtract, multiply, divide, power, sqrt, factorial)
                a: First number
                b: Second number (not used for sqrt and factorial)
                precision: Number of decimal places for result
            
            Returns:
                The result of the calculation
            """
            try:
                if operation == "add":
                    result = a + b
                elif operation == "subtract":
                    result = a - b
                elif operation == "multiply":
                    result = a * b
                elif operation == "divide":
                    if b == 0:
                        raise ValueError("Division by zero is not allowed")
                    result = a / b
                elif operation == "power":
                    result = a ** b
                elif operation == "sqrt":
                    if a < 0:
                        raise ValueError("Square root of negative number is not allowed")
                    result = a ** 0.5
                elif operation == "factorial":
                    if a < 0 or a != int(a):
                        raise ValueError("Factorial is only defined for non-negative integers")
                    result = 1
                    for i in range(1, int(a) + 1):
                        result *= i
                else:
                    raise ValueError(f"Unsupported operation: {operation}")
                
                return f"{result:.{precision}f}" if isinstance(result, float) else str(result)
                
            except Exception as e:
                return f"Error: {str(e)}"
        
        @self.server.tool("server_stats")
        async def server_stats() -> str:
            """
            Get server statistics and health information.
            
            Returns:
                Server statistics in JSON format
            """
            stats = {
                "server_name": "calculator-server",
                "version": "1.0.0",
                "uptime_seconds": time.time() - getattr(self, '_start_time', time.time()),
                "is_running": self.is_running,
                "transport_protocol": "tcp",
                "transport_port": 8080,
                "tools_registered": ["calculator", "server_stats"],
                "memory_usage": "N/A",  # Would need psutil for real memory stats
                "cpu_usage": "N/A",     # Would need psutil for real CPU stats
                "timestamp": time.time()
            }
            return json.dumps(stats, indent=2)
        
        print("✅ Calculator tools registered: calculator, server_stats")
    
    def _create_transport_config(self) -> GopherTransportConfig:
        """Create transport configuration with comprehensive filters."""
        return GopherTransportConfig(
            name="calculator-server-transport",
            protocol="tcp",
            host="localhost",
            port=8080,
            timeout=30000,
            max_connections=100,
            buffer_size=16384,
            session_timeout=3600000,  # 1 hour
            max_sessions=1000,
            tls_enabled=False,
            keep_alive=True,
            keep_alive_interval=30000,
            max_idle_time=300000,
            filters=FilterManagerConfig(
                # Security filters
                security=SecurityFilterConfig(
                    authentication=AuthenticationConfig(
                        enabled=True,
                        method="jwt",
                        secret="calculator-server-secret-key",
                        issuer="calculator-server",
                        audience="calculator-clients",
                        algorithms=["HS256", "RS256"]
                    ),
                    authorization=AuthorizationConfig(
                        enabled=True,
                        policy="allow",
                        rules=[
                            {"resource": "/tools/calculator", "action": "call", "role": "user"},
                            {"resource": "/tools/server_stats", "action": "call", "role": "user"},
                            {"resource": "/tools/*", "action": "list", "role": "user"}
                        ],
                        default_action="deny"
                    )
                ),
                
                # Observability filters
                observability=ObservabilityFilterConfig(
                    access_log=AccessLogConfig(
                        enabled=True,
                        format="json",
                        include_headers=True,
                        include_body=False,
                        max_body_size=4096
                    ),
                    metrics=MetricsConfig(
                        enabled=True,
                        namespace="calculator_server",
                        labels={
                            "server": "calculator-server",
                            "version": "1.0.0",
                            "environment": "demo"
                        },
                        histogram_buckets=[0.1, 0.5, 1.0, 2.5, 5.0, 10.0, 25.0, 50.0, 100.0]
                    ),
                    tracing=TracingConfig(
                        enabled=True,
                        service_name="calculator-server-gopher",
                        sampler_type="const",
                        sampler_param=1.0,
                        headers=["x-trace-id", "x-span-id", "x-b3-traceid", "x-b3-spanid"]
                    )
                ),
                
                # Traffic management filters
                traffic_management=TrafficManagementFilterConfig(
                    rate_limit=RateLimitConfig(
                        enabled=True,
                        requests_per_minute=2000,  # Higher rate limit for server
                        burst_size=200,
                        key_extractor="ip"
                    ),
                    circuit_breaker=CircuitBreakerConfig(
                        enabled=True,
                        failure_threshold=10,  # Higher threshold for server
                        recovery_timeout=60000,
                        half_open_max_calls=5,
                        slow_call_threshold=5000
                    ),
                    retry=RetryConfig(
                        enabled=True,
                        max_attempts=3,
                        initial_delay=1000,
                        max_delay=10000,
                        backoff_multiplier=2.0,
                        retryable_status_codes=[500, 502, 503, 504, 408, 429]
                    ),
                    load_balancer=LoadBalancerConfig(
                        enabled=True,
                        strategy="round_robin",
                        upstreams=[
                            {"host": "backend1.example.com", "port": 8080, "weight": 1, "health_check": True},
                            {"host": "backend2.example.com", "port": 8080, "weight": 1, "health_check": True}
                        ],
                        health_check_interval=30000,
                        health_check_timeout=5000
                    )
                ),
                
                # HTTP filters
                http=HttpFilterConfig(
                    codec=HttpCodecConfig(
                        enabled=True,
                        version="1.1",
                        max_header_size=8192,
                        max_body_size=1048576,
                        chunked_encoding=True
                    ),
                    router=HttpRouterConfig(
                        enabled=True,
                        routes=[
                            {"path": "/api/v1", "target": "backend-v1", "methods": ["GET", "POST"]},
                            {"path": "/api/v2", "target": "backend-v2", "methods": ["GET", "POST", "PUT", "DELETE"]}
                        ],
                        default_route="default-backend",
                        strip_prefix=False
                    ),
                    compression=HttpCompressionConfig(
                        enabled=True,
                        algorithms=["gzip", "deflate", "brotli"],
                        min_size=1024,
                        level=6
                    )
                ),
                
                # Error handling
                error_handling=ErrorHandlingConfig(
                    stop_on_error=False,
                    retry_attempts=3,
                    fallback_behavior=FallbackBehavior.PASSTHROUGH
                )
            ),
            
            # CApiFilter integration
            custom_callbacks=self._create_custom_callbacks()
        )
    
    def _create_custom_callbacks(self) -> Dict[str, Any]:
        """Create custom callbacks for CApiFilter integration."""
        def on_message_received(buf, end_stream, user_data):
            """Callback for when a message is received."""
            print(f"🔍 [CApiFilter DEBUG] onMessageReceived callback called! Buffer: {buf}, EndStream: {end_stream}")
            return 0  # MCP_FILTER_CONTINUE
        
        def on_message_sent(buf, end_stream, user_data):
            """Callback for when a message is sent."""
            print(f"🔍 [CApiFilter DEBUG] onMessageSent callback called! Buffer: {buf}, EndStream: {end_stream}")
            return 0  # MCP_FILTER_CONTINUE
        
        def on_connection_established(user_data, fd):
            """Callback for when a connection is established."""
            print(f"🔍 [CApiFilter DEBUG] onConnectionEstablished callback called! FD: {fd}")
        
        def on_high_watermark(user_data):
            """Callback for high watermark."""
            print(f"🔍 [CApiFilter DEBUG] onHighWatermark callback called!")
        
        def on_low_watermark(user_data):
            """Callback for low watermark."""
            print(f"🔍 [CApiFilter DEBUG] onLowWatermark callback called!")
        
        def on_error(user_data, code, msg):
            """Callback for errors."""
            message = msg.value.decode('utf-8') if msg else "Unknown error"
            print(f"🔍 [CApiFilter DEBUG] onError callback called! Code: {code}, Message: {message}")
        
        return {
            "on_data": on_message_received,
            "on_write": on_message_sent,
            "on_new_connection": on_connection_established,
            "on_high_watermark": on_high_watermark,
            "on_low_watermark": on_low_watermark,
            "on_error": on_error,
            "user_data": None,
        }
    
    async def stop(self):
        """Stop the calculator server."""
        print("🛑 Stopping calculator server...")
        
        self.is_running = False
        
        if self.transport:
            await self.transport.close()
            print("✅ Transport closed")
        
        print("✅ Calculator server stopped")
    
    async def run_forever(self):
        """Run the server forever."""
        self._start_time = time.time()
        
        # Set up signal handlers for graceful shutdown
        def signal_handler(signum, frame):
            print(f"\n📡 Received signal {signum}, shutting down gracefully...")
            asyncio.create_task(self.stop())
        
        signal.signal(signal.SIGINT, signal_handler)
        signal.signal(signal.SIGTERM, signal_handler)
        
        try:
            # Keep the server running
            while self.is_running:
                await asyncio.sleep(1)
        except KeyboardInterrupt:
            print("\n📡 Received keyboard interrupt, shutting down...")
        finally:
            await self.stop()


async def main():
    """Main entry point for the calculator server."""
    print("🧮 MCP Calculator Server with GopherTransport")
    print("=" * 50)
    
    server = CalculatorServer()
    
    try:
        await server.start()
        await server.run_forever()
    except Exception as e:
        print(f"❌ Server error: {e}")
        import traceback
        traceback.print_exc()
    finally:
        await server.stop()


if __name__ == "__main__":
    asyncio.run(main())
